Building with libstdc++ 8 needs separate library providing std::filesystem?

## Description
Building the latest version on a linux workstation (RHEL 8) with libstdc++ 8.5 fails during linking.

## Shortened error message 
```
undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
```
This appears to be related to `std::filesystem` being included in `stdc++fs` (rather than in stdc++ itself) for old versions of gcc. As far as I know, this is not linked ordinarily, as it is not required by gcc > 8.

## EDIT: Potential fix
Adding the following to CMakeLists.txt fixes this on my system (shamelessly taken from a post on cmake.org)

```cmake
function( set_required_build_settings_for_GCC8 )
    # link with libstdc++fs.a when using GCC 8
    link_libraries( "$<$<AND:$<CXX_COMPILER_ID:GNU>,$<VERSION_LESS:$<CXX_COMPILER_VERSION>,9.0>>:-lstdc++fs>" )
endfunction()
set_required_build_settings_for_GCC8()
```

## System information
Please check all relevant boxes and provide details.

- Operating system:
  - Linux: RHEL 8, libstdc++ 8.5
- Version of WarpX: latest
- Installation method:
  - From source with CMake: 
  - cmake -S . -B build -DWarpX_COMPUTE=OMP -DWarpX_DIMS="2" -DWarpX_QED=OFF -DWarpX_EB=OFF

## Full error message

<details>
[100%] Linking CXX executable bin/warpx.2d.MPI.OMP.DP.PDP.OPMD
lib/libwarpx.2d.MPI.OMP.DP.PDP.OPMD.a(SemiImplicitEM.cpp.o): In function `PicardSolver<WarpXSolverVec, ImplicitSolver>::Define(WarpXSolverVec const&, ImplicitSolver*)':
SemiImplicitEM.cpp:(.text._ZN12PicardSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_[_ZN12PicardSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_]+0x120): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
SemiImplicitEM.cpp:(.text._ZN12PicardSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_[_ZN12PicardSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_]+0x130): undefined reference to `std::filesystem::__cxx11::path::parent_path() const'
SemiImplicitEM.cpp:(.text._ZN12PicardSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_[_ZN12PicardSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_]+0x140): undefined reference to `std::filesystem::create_directories(std::filesystem::__cxx11::path const&)'
lib/libwarpx.2d.MPI.OMP.DP.PDP.OPMD.a(SemiImplicitEM.cpp.o): In function `NewtonSolver<WarpXSolverVec, ImplicitSolver>::Define(WarpXSolverVec const&, ImplicitSolver*)':
SemiImplicitEM.cpp:(.text._ZN12NewtonSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_[_ZN12NewtonSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_]+0x8a2): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
SemiImplicitEM.cpp:(.text._ZN12NewtonSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_[_ZN12NewtonSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_]+0x8b5): undefined reference to `std::filesystem::__cxx11::path::parent_path() const'
SemiImplicitEM.cpp:(.text._ZN12NewtonSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_[_ZN12NewtonSolverI14WarpXSolverVec14ImplicitSolverE6DefineERKS0_PS1_]+0xb64): undefined reference to `std::filesystem::create_directories(std::filesystem::__cxx11::path const&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/app_2d.dir/build.make:111: bin/warpx.2d.MPI.OMP.DP.PDP.OPMD] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:3671: CMakeFiles/app_2d.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
</details>

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -206,6 +206,17 @@ include(CTest)
 # Dependencies ################################################################
 #
 
+# Old GCCs: std::filesystem as a separate library
+try_compile(CXX_HAS_STD_FS "${WarpX_BINARY_DIR}/temp"
+    "${WarpX_SOURCE_DIR}/cmake/has_filesystem.cpp"
+    CMAKE_FLAGS -DCMAKE_CXX_STANDARD=17 -DCMAKE_CXX_STANDARD_REQUIRED=ON
+)
+try_compile(CXX_HAS_STD_FS_WITH_LINK "${WarpX_BINARY_DIR}/temp"
+    "${WarpX_SOURCE_DIR}/cmake/has_filesystem.cpp"
+    CMAKE_FLAGS -DCMAKE_CXX_STANDARD=17 -DCMAKE_CXX_STANDARD_REQUIRED=ON
+    LINK_LIBRARIES stdc++fs
+)
+
 # AMReX
 #   builds AMReX from source (default) or finds an existing install
 set(WarpX_amrex_dim ${WarpX_DIMS})  # RZ is AMReX 2D
@@ -293,6 +304,16 @@ foreach(D IN LISTS WarpX_DIMS)
                 # Number must be more than the number of .cpp files in WarpX
                 UNITY_BUILD_BATCH_SIZE 10000
             )
+        endif()
+
+        # GCC < 9 need to manually link std::filesystem
+        if(NOT CXX_HAS_STD_FS)
+            if(CXX_HAS_STD_FS_WITH_LINK)
+                target_link_libraries(ablastr_${SD} PUBLIC stdc++fs)
+            else()
+                message(FATAL_ERROR "Your compiler does not support C++17 "
+                        "<filesystem> Please use a newer C++ compiler.")
+            endif()
         endif()
     endif()
 

--- a/cmake/has_filesystem.cpp
+++ b/cmake/has_filesystem.cpp
@@ -1,0 +1,12 @@
+#include <filesystem>
+
+
+int main()
+{
+    namespace fs = std::filesystem;
+
+    fs::path dir{"./mkdir_test"};
+    fs::create_directories(dir);
+
+    return 0;
+}

